### PR TITLE
[SOLID] Change if && to early return

### PR DESCRIPTION
**Pull request by dobryy**

### CODE REVIEW

**Comment by dobryy** [comment]
Just checking out very basic logic...

File: rules/nette/src/TemplatePropertyAssignCollector.php [hunk 1]
```diff
@@ -100,11 +100,13 @@ private function collectVariableFromAssign(Assign $assign): void
 
             $this->nodesToRemove[] = $assign;
         }
-
-        // $x = $this->template
```
Comment (TomasVotruba) [inline rules/nette/src/TemplatePropertyAssignCollector.php:104]: This comment should be kept

File: rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        $invertedRightCondition = $this->conditionInverter->createInvertedCondition($expr->right);
+
+        $firstIf = new If_($invertedLeftCondition);
+        $firstIf->stmts = [new Return_()];
+        $secondIf = new If_($invertedRightCondition);
+        $secondIf->stmts = [new Return_()];
+
+        $this->addNodeAfterNode($firstIf, $node);
+        $this->addNodeAfterNode($secondIf, $node);
+        $this->addNodeAfterNode($ifReturn, $node);
```
Comment (TomasVotruba) [inline rules/solid/src/Rector/If_/ChangeAndIfToEarlyReturnRector.php:125]: These is also `addNodesAfterNode()` method

File: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace Rector\SOLID\Tests\Rector\If_\ChangeAndIfToEarlyReturnRector\Fixture;
+
+class EqualsClass
+{
+    public function canDrive(Car $car)
+    {
+        if ($car->hasWheels === true && $car->hasFuel === true) {
```
Comment (TomasVotruba) [inline rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc:9]: Could you also add test for 3 items?

```php
$value === true && $value === true && $value === true 
```

Comment (dobryy) [reply rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc:9]: At the moment, the case with 3 items is skipped, there is a fixture for this. More than two items will make rule a bit more complicated I think. I wanted to keep the PR as small as possible.

File: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_more_than_two_conditions.php.inc [hunk 4]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+namespace Rector\SOLID\Tests\Rector\If_\ChangeAndIfToEarlyReturnRector\Fixture;
+
+class SkipMoreThanTwoConditionClass
+{
+    public function canDrive(Car $car)
+    {
+        if ($car->hasWheels && $car->hasFuel && $car->hasEngine) {
```
Comment (dobryy) [inline rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_more_than_two_conditions.php.inc:9]: @TomasVotruba This situation can be handled in the next PR.

File: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_non_first_level_if.php.inc [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    {
+        if ($car->wheelsCount() > 2) {
+            if ($car->hasWheels && $car->hasFuel) {
+                $this->canDrive = true;
+            }
+
+            return;
+        }
+
+        $this->canDrive = true;
```
Comment (dobryy) [inline rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_non_first_level_if.php.inc:17]: @TomasVotruba This situation can be handled in the next PR.

File: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_if_in_foreach.php.inc [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    private function refactorStmts(array $nodes): void
+    {
+        foreach ($nodes as $node) {
+            if ($node instanceof Name && $node instanceof Identifier) {
+                $this->processNameOrIdentifier($node);
+            }
+
+            return;
+        }
+
```
Comment (dobryy) [inline rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/skip_if_in_foreach.php.inc:19]: @TomasVotruba This situation looks crazy but also can be handled in the next PR.

File: rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc (hunk 3 above)
Comment (TomasVotruba) [reply rules/solid/tests/Rector/If_/ChangeAndIfToEarlyReturnRector/Fixture/equals.php.inc:9]: I see

**Review by TomasVotruba — CHANGES_REQUESTED** [review]
I've added few comments

**Comment by dobryy** [comment]
> I've added few comments

I've implemented changes as a new commit https://github.com/rectorphp/rector/pull/4344/commits/38047f43579ff5916c14f97fbd122002bb0464b9

**Comment by dobryy** [comment]
@TomasVotruba Pipeline is finally green.

**Comment by TomasVotruba** [comment]
Thank you

**Comment by ERuban** [comment]
@dobryy @TomasVotruba hey, I've just a one question - why it is a SOLID rule?

**Comment by TomasVotruba** [comment]
From all the sets there are now, SOLID and early return are closest.
What set would you suggest instead?

**Comment by ERuban** [comment]
@TomasVotruba So, I would suggest `CodingStyle` set for this rule - seems it will be better here, just because it is really more about accepted conventions in teams/companies, imho.
Also, why do you think it is closest to SOLID rules?

**Comment by ERuban** [comment]
Hm, it will be really nice to move it to a standalone set. 
Or, maybe, to CodeQuality set :)

**Comment by TomasVotruba** [comment]
Then, next month we a question, why it's part of CodeQuality set :D 

Btw, is current situation causing any undesired behavior of Rector in your code or is it rather academic curiosity?

**Comment by TomasVotruba** [comment]
There is an issue here: https://github.com/rectorphp/rector/issues/4464

**Comment by ERuban** [comment]
So, ok, standalone set looks ok, imo. 
Yeap, it is just academic curiosity (we haven't such rule in our team). 

Btw, I've some issues after updating Rector from the older version, but will ask you separately about it. ;)

**Pull request merged**